rust/src/container.rs

```diff
@@ -0,0 +1,104 @@
+//! CLI exposing `ostree-rs-ext container`
+
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use std::convert::TryInto;
+
+use anyhow::{Context, Result};
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+enum Opts {
+    /// Import an ostree commit embedded in a remote container image
+    Import {
+        /// Path to the repository
+        #[structopt(long)]
+        repo: String,
+
+        /// Image reference, e.g. registry:quay.io/exampleos/exampleos:latest
+        imgref: String,
+    },
+
+    /// Print information about an exported ostree-container image.
+    Info {
+        /// Image reference, e.g. registry:quay.io/exampleos/exampleos:latest
+        imgref: String,
+    },
+
+    /// Export an ostree commit to an OCI layout
+    Export {
+        /// Path to the repository
+        #[structopt(long)]
+        repo: String,
+
+        /// The ostree ref or commit to export
+        rev: String,
+
+        /// Image reference, e.g. registry:quay.io/exampleos/exampleos:latest
+        imgref: String,
+    },
+}
+
+async fn container_import(repo: &str, imgref: &str) -> Result<()> {
+    let repo = &ostree::Repo::open_at(libc::AT_FDCWD, repo, gio::NONE_CANCELLABLE)?;
```

Comment:
I have the feeling that we are starting to go down the path of freely mixing async and blocking logic, which is... not very kosher.
This is likely not a problem here (single-purpose CLI, multi-threaded runtime), but doing the same further into library code will eventually come and bite us.
I wish we had a better way upfront to scaffold and split apart colored functions, but I don't have a good alternative to offer.

Reply:
I'm not super concerned about counting random one-off local file I/O as blocking, particularly things that are almost certain to be cached already.

I think local file I/O is more concerning when at least 2 of these are true:

- The process is also exposing a "service" itself (e.g. HTTP or DBus)
- There's a *lot* of local file I/O, e.g. it's more like a database than just opening a config file
- The files are unlikely to be cached already (but here the ostree repo directory is almost certain to already live in memory because the rpm-ostree daemon is using inotify on it, etc.)

In this case, rpm-ostreed is a service *but* we'd likely run this in a worker thread anyways because that's how the rest of the code works.  So none of the 3 really apply.

Reply:
(This code of course does do a lot of file I/O but all that happens inside a `spawn_blocking()` worker thread (which would be distinct from a higher level worker thread used by rpm-ostree (threads! we love them! (also parentheticals))))